custom_components/deebot/config_flow.py

```diff
@@ -0,0 +1,144 @@
+    """Validate the user input allows us to connect.
+    Data has the keys from DATA_SCHEMA with values provided by the user.
+    """
+
+    if len(data[CONF_COUNTRY]) != 2:
+        raise InvalidCountry
+
+    if len(data[CONF_CONTINENT]) != 2:
+        raise InvalidContinent
+
+    return await hass.async_add_executor_job(ConfigEntryRetriveRobots, hass, data)
+
+
+def ConfigEntryRetriveRobots(hass: core.HomeAssistant, domain_config):
+    ecovacs_api = EcoVacsAPI(
+        DEEBOT_API_DEVICEID,
+        domain_config.get(CONF_USERNAME),
+        EcoVacsAPI.md5(domain_config.get(CONF_PASSWORD)),
+        domain_config.get(CONF_COUNTRY),
+        domain_config.get(CONF_CONTINENT),
+    )
+
+    return ecovacs_api.devices()
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Deebot."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_CLOUD_POLL
+
+    async def async_step_user(self, user_input=None):
+        """Handle the initial step."""
+        self.data = {}
+        errors = {}
+        if user_input is not None:
+            try:
+                info = await validate_input(self.hass, user_input)
+                self.robot_list = info
+            except CannotConnect:
+                errors["base"] = "cannot_connect"
+            except InvalidCountry:
+                errors[CONF_COUNTRY] = "invalid_country"
+            except InvalidContinent:
+                errors[CONF_CONTINENT] = "invalid_continent"
+            except Exception:
+                _LOGGER.exception("Unexpected exception")
+                errors["base"] = "unknown"
+
+            if not errors:
+                self.data = user_input
+
+                robot_listDict = {e["name"]: e["nick"] for e in self.robot_list}
+                options_schema = vol.Schema(
+                    {
+                        vol.Required(
+                            CONF_DEVICEID, default=list(robot_listDict.keys())
+                        ): cv.multi_select(robot_listDict)
+                    }
+                )
+
+                return self.async_show_form(
+                    step_id="robots", data_schema=options_schema, errors=errors
+                )
+
+        # If there is no user input or there were errors, show the form again, including any errors that were found with the input.
+        return self.async_show_form(
+            step_id="user", data_schema=DATA_SCHEMA, errors=errors
+        )
+
+    async def async_step_robots(self, user_input=None):
+        """Handle the robots selection step."""
+
+        errors = {}
+        if user_input is not None:
+            try:
+                if len(user_input[CONF_DEVICEID]) < 1:
+                    errors["base"] = "select_robots"
+                else:
+                    self.data[CONF_DEVICEID] = user_input
+                    return self.async_create_entry(
+                        title="Ecovacs vacuum custom integration", data=self.data
```

Comment:
I suggest that we set as title the name of the vacuum. So the user can distinguish the vacuums on the integration side. Currently we create each entry with the same name. For example compare it with EspHome in the screenshot below (esp01, esp02, ... are the device names. I personally called them in this manner)
![grafik](https://user-images.githubusercontent.com/26537646/110693281-8ee91300-81e7-11eb-8f86-9236b9a11b61.png)
To get the title correct, we probably should only allow single select on the robots list or create of each selected robot a different entry.


Currently it is also possible to add the same vacuum twice. I think we should throw an error, when the vacuum is already set up.

Reply:
okay I think I have a idea for the problem. Will give it a try tomorrow

Reply:
I don't know, i would like have it by account.. insert multiple times same user and password for different robots on same account is kinda confusing

Reply:
The account is fine for me. Only we should add a OptionHandler, so a user can change room color and co without deleting and creating the entry again

Reply:
Yes, i agree with that, maybe use first "form" to insert only Username,Password,Country,Continent -> Select robots and then from Options you should choose live map and room colors.. maybe add more options such as map color, line color etc..

Reply:
At the moment im honestly a bit scared from ecovacs api.. it seems like they started changes endpoints. i hope they dont change it too much

Reply:
> At the moment im honestly a bit scared from ecovacs api.. it seems like they started changes endpoints. i hope they dont change it too much

Not good et all.... Next week I hopefully set up [bumper](https://github.com/bmartin5692/bumper) (selfhosted deebot server) so my vacuum is not anymore sending data to china

Reply:
@And3rsL Since Version 2 of the app ecovacs uses oauth. Probably that was the big api change.
Give a look my PR at bumper https://github.com/bmartin5692/bumper/pull/116
There exists new endpoints like `/newauth.do` and `api/appsvr/oauth_callback` which we must probably use in the future.
After calling the oauth_callback each post request has the following in it's body
```json
"auth": {
        "realm": "ecouser.net",
        "resource": "IASB481A03G8",
        "token": "TOKEN",
        "userid": "USERID",
        "with": "users"
    }
```